File: BalancedBinaryTree/third.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

int checkBalanceAndDepth(struct TreeNode* node, bool *isBalanced)
{
    int leftDepth = node->left == NULL? 0 : checkBalanceAndDepth(node->left, isBalanced);
    if(!*isBalanced)
    {
        return -1;
    }
    int rightDepth = node->right == NULL? 0 :checkBalanceAndDepth(node->right, isBalanced);
    if(!*isBalanced)
    {
        return -1;
    }
    int diff = leftDepth - rightDepth;
    *isBalanced = (diff == -1 || diff == 0 || diff == 1);
    return leftDepth > rightDepth? leftDepth + 1 : rightDepth + 1;
}
bool isBalanced(struct TreeNode* root) {
    if(root == NULL) return true;
    bool balanced = true;
    checkBalanceAndDepth(root, &balanced);
    return balanced;
}

File: BalancedBinaryTree/third.py
cal the depth from the leaf to a tree node, judge if it is balanced and store it in each node.

class Solution:
    # @param {TreeNode} root
    # @return {boolean}
    def isBalanced(self, root):
        if root == None: return True
        stack = []
        p = root
        while p != None or len(stack) > 0:
            while p != None:
                    stack.append((0,p))
                    p = p.left
            lr, p = stack.pop()
            if p.right != None:
                if lr == 0:
                    stack.append((1,p))
                    p = p.right
                    continue
                r = p.right.val
            else:
                r = 0
            if p.left == None:
                l = 0
            else:
                l = p.left.val

            if (r-l) in range(-1,2):
                p.val = max(r,l)+1
                p = None
            else:
                return False

        return True
